scapy/layers/l2.py

```diff
@@ -38,39 +38,62 @@
 from scapy.sendrecv import sendp, srp, srp1
 from scapy.utils import checksum, hexdump, hexstr, inet_ntoa, inet_aton, \
     mac2str, valid_mac, valid_net, valid_net6
+from scapy.base_classes import Packet_metaclass
+from scapy.compat import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
+from scapy.interfaces import NetworkInterface
 if conf.route is None:
     # unused import, only to initialize conf.route
     import scapy.route  # noqa: F401
 
 
+# type definitions
+ResolverCallable = Callable[[Packet_metaclass, Packet_metaclass],
+                            Optional[str]]
+
 #################
 #  Tools        #
 #################
 
 
 class Neighbor:
     def __init__(self):
-        self.resolvers = {}
+        # type: () -> None
+        self.resolvers = {}  # type: Dict[Tuple[Packet_metaclass, Packet_metaclass], ResolverCallable] # noqa: E501
 
     def register_l3(self, l2, l3, resolve_method):
+        # type: (Packet_metaclass, Packet_metaclass, ResolverCallable) -> None
         self.resolvers[l2, l3] = resolve_method
 
     def resolve(self, l2inst, l3inst):
+        # type: (Ether, Packet) -> Optional[str]
         k = l2inst.__class__, l3inst.__class__
         if k in self.resolvers:
             return self.resolvers[k](l2inst, l3inst)
+        return None
 
     def __repr__(self):
+        # type: () -> str
         return "\n".join("%-15s -> %-15s" % (l2.__name__, l3.__name__) for l2, l3 in self.resolvers)  # noqa: E501
 
 
-conf.neighbor = Neighbor()
+neighbor = Neighbor()
```

Comment:
I understand why you did that, but this is going to break backward compatibility.
Could you at least add `conf.neighbor = neighbor` afterwards?

Reply:
This still causes the mypy error:

    scapy/layers/l2.py:89: error: "Conf" has no attribute "neighbor"

I saw later that `Conf` has another way to solve this issue for `conf.route`:

    route = None  # type: 'scapy.route.Route'

Although I'm not sure how it works. Shouldn't the type be an Optional if it is initialized to None?

Reply:
That's actually a good idea, could you do the same than for `route` ?

> Shouldn't the type be an Optional if it is initialized to None?

It should, but I think mypy doesn't check that when using absolute paths (such as here). Even if it did check, we would probably just ignore this case. I'd like to have the conf attributes typed as if they were never None when they are initialized on import.

Reply:
> That's actually a good idea, could you do the same than for `route` ?

You mean for `neighbor`? Sure, I'll try. `route` is already implemented this way.

Reply:
That's what I meant yes, thanks !